### ISSUE TITLE
Option to allow pretty print without alpha sorting

### ISSUE THREAD

**Issue by patcon** [jakubroztocil/httpie#128]
Right now, the options are `colors`, `format`, and `all`, but often the order of the json returned is intentional and for clarify. It would be nice to be able to pretty print the json without having it's keys sorted :)

Awesome tool, and thanks very much for everything!

**Comment by pontuspalmenas** [comment]
+1
Please add this option. Sorting is a deal breaker for me.

**Comment by rawaludin** [comment]
Any news about this?

**Comment by justCxx** [comment]
+1

**Comment by tedmiston** [comment]
+1

**Comment by hartym** [comment]
This problem is indeed misleading. I struggled a bit to understand why my (server side) OrderedDict was not correctly serialised in my app, then I finally understood that the culprit was not the server but indeed the client re-sorting the keys. It is sometimes useful to have the keys sorted, but for an HTTP client, having a default that changes the response content does not sound like the right thing.

Thanks for the amazing tool anyway ;)

**Comment by msabramo** [comment]
@jkbrzt: You mention a while back in https://github.com/jkbrzt/httpie/pull/151#issuecomment-139560575 that you wanted a more generic solution to this problem; was any progress made on that?

**Comment by jkbrzt** [comment]
@msabramo it's still on my to do list but not progress yet

**Comment by NN88** [comment]
+ 1 this really needs to become a feature. 

Thanks for this amazing tool!

**Comment by kylebebak** [comment]
I wrote an HTTP client for Sublime Text that's also built on top of Requests, called Requester. I also wanted to allow users to choose whether or not keys for JSON responses are sorted, [so I added a 'fmt' argument to calls to requests](https://github.com/kylebebak/Requester#json-response-formatting).

[I solved it like this](https://github.com/kylebebak/Requester/blob/e7c0ddf82ad377375c3c18b0918ee3fecb9815a9/commands/request.py#L44). If the response body is JSON, calling `res.json()` on the response object `res` returns a Python dictionary. At this point you can pass the dict to `json.dumps` to specify indentation and whether keys are sorted. However, if you just return `res.text` (equivalent to returning the raw response body) the keys won't be sorted.

**Comment by opensas** [comment]
Has this feature been added? I tried with `--pretty=unsorted` but I get

```
http: error: argument --pretty: invalid choice: 'unsorted' (choose from 'all', 'colors', 'format', 'none')
```

Anyway, just passing `--pretty=colors` did the trick, the output is not sorted (I already format the output), but I found no wayt to specify something like `--pretty=colors,format`

```bash
$ http --version
0.9.9
```

**Comment by virgo47** [comment]
This feature should be default. I love HTTPie, but resorting the keys - while convenient for some usage - is confusing if you don't realize, that client tool does this. Piping `http` to `jq -C` is a workaround, but it should be default behavior. While the order is not probably that important for a machine, it may be for a human observer.

**Comment by nmtitov** [comment]
Knock-knock... 7 years after... any updates?

**Comment by opensas** [comment]
Anybody know of an httpie fork that took care of this annoying issue?

**Comment by opensas** [comment]
I found the following workaround, instead of using httpie I found [curlie](https://curlie.io/), which seems to

> Like curl but unlike httpie, headers are written on stderr instead of stdout.

So I can use it like this:

```shell
$ curlie GET localhost:3000/tasks | jq -C
HTTP/1.1 200 OK
X-Powered-By: Express
Content-Type: application/json; charset=utf-8
Content-Length: 304
ETag: W/"130-ED1W4hQo1i7na7wy5Ewc7iKdoJc"
Date: Wed, 27 May 2020 06:28:26 GMT
Connection: keep-alive

[
  {
    "id": 2,
    "title": "new task2",
    "description": "description2",
    "status": "OPEN"
  },
  {
    "id": 3,
    "title": "new task3",
    "description": "description3",
    "status": "OPEN"
  }
]
```

Where as with httpie I would the headers

BTW, I created this convenient script:

```shell
$ cat ~/bin/c
curlie "$@" | jq -C
```

**Comment by jakubroztocil** [comment]
How about this? 


Disable the default header and JSON key sorting, and specify a custom JSON indent size:

```bash
$ http --format-options headers.sort=false,json.sort_keys=false,json.indent=2 httpbin.org/get
```

```
  --format-options FORMAT_OPTIONS
      Controls output formatting. Only relevant when formatting is enabled
      through (explicit or implied) --pretty=all or --pretty=format.
      The following are the default options:

          headers.sort=true
          json.format=true
          json.indent=4
          json.sort_keys=true

      You can specify multiple comma-separated options. For example, this modifies
      the settings to disable the sorting of headers and JSON keys:

          --format-options headers.sort=false,json.sort_keys=false

      This is something you will typically put into your config file.
```

It’s already `master` & feedback would be appreciated! ([Install HTTPie from master](https://httpie.org/docs#unstable-version).)

**Comment by jakubroztocil** [comment]
The idea behind the syntax is to allow more options to be easily added in the future. But I do realise that it’s quite a verbose one. On the other hand, I would expect people to set it once in the config file and not type it manually. Thoughts?

**Comment by opensas** [comment]
Thanks a lot for the reply, being able to configure those options for format output would be great, speacially if you can just put it on a conf file.

I tried to install from master but I have the following issue:

```
$ sudo apt install python-pip
[...]

$ pip install --upgrade https://github.com/jakubroztocil/httpie/archive/master.tar.gz
Collecting https://github.com/jakubroztocil/httpie/archive/master.tar.gz
  Downloading https://github.com/jakubroztocil/httpie/archive/master.tar.gz
     | 2.2MB 34.5MB/s
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-9g7C_j-build/setup.py", line 74
        download_url=f'https://github.com/jakubroztocil/httpie/archive/{httpie.__version__}.tar.gz',
                                                                                                  ^
    SyntaxError: invalid syntax
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-9g7C_j-build/

$ pip --version
pip 9.0.1 from /usr/lib/python2.7/dist-packages (python 2.7)

$ python --version
Python 2.7.17

$ uname -a
Linux xps 5.0.0-1052-oem-osp1 #57-Ubuntu SMP Mon May 4 11:39:08 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

Do I need python 3?

**Comment by jakubroztocil** [comment]
@opensas thanks for testing it out! Yes, you need Python 3.6+ (Python 2.7 is dead, and since v2.0.0 HTTPie does not support it).

**Comment by jakubroztocil** [comment]
@opensas if you have Docker installed, you can test it in a container:

```bash
$ docker run --rm -it python:3.8-slim
```

Then inside container:

```bash
$ pip install --upgrade https://github.com/jakubroztocil/httpie/archive/master.tar.gz
```

```bash
$ http --version
2.2.0-dev
```

**Comment by opensas** [comment]
I could install it with `python3 -m pip install --upgrade https://github.com/jakubroztocil/httpie/archive/master.tar.gz`

I tested it and it works ok with `http --format-options headers.sort=false,json.sort_keys=false,json.indent=2`

I'm trying to set those options in my config file, but I guess I have a trouble with the xx=xx options, I'm trying like this:

```
$ cat ~/.config/httpie/config.json
{
    "default_options": [
      "--format-options=headers.sort=false,json.sort_keys=false,json.indent=2"
    ]
}
```
What should be the correct syntax for my config.json file? DO I have to escape the `=` signs?

**Comment by opensas** [comment]
ok, I tried copying exactly the example provided in the [docs](https://github.com/jakubroztocil/httpie#default-options) but it's not taking the changes in the `~/.config/httpie/config.json` file

```
$ cat ~/.config/httpie/config.json

{
    "default_options": [
      "--style=fruity"
    ]
}
```

**Comment by jakubroztocil** [comment]
@opensas what output do you get when you run `http --debug`?


![image](https://user-images.githubusercontent.com/326885/83037726-6e771000-a03c-11ea-8fe2-1eb94c916ca2.png)

**Comment by jakubroztocil** [comment]
@opensas the syntax is correct (it should be the same like on the CLI, there nothing special about the config options).

**Comment by opensas** [comment]
Sorry, I found the issue, I had a `~/.httpie/config.json` from previous installation, so it was bypassing the file I created at `~/.config/httpie/config.json`

just removed the .httpie folder and now everything works as expected

thanks a lot! I hope you can release this changes in the near future!

Some feedback:

- I would consider setting headers.sort and json.sort to false by default, I think that changing the output order should be the exception and not the rule

- perhaps also add a --format-options sort=[true|false] shortcut to set both headers.sort and json.sort at the same time

anyway, congrats on the excellent job you are doing

**Comment by opensas** [comment]
more feedback:

I think the easiest thing to use from the command line would be to allow something like:

```
--pretty=colors,format,unsorted

--pretty=all,unsorted
```

**Comment by jakubroztocil** [comment]
> Sorry, I found the issue, I had a `~/.httpie/config.json` from previous installation, so it was bypassing the file I created at `~/.config/httpie/config.json` 
> just removed the .httpie folder and now everything works as expected

The `$XDG_CONFIG_HOME`-based  `~/.config/httpie` location is also new in the yet-to-be-released version. Looks like a warning if both the locations exist would be helpful.

> thanks a lot! I hope you can release this changes in the near future!

Yes, it should be out pretty soon.

> I would consider setting `headers.sort` and `json.sort` to `false` by default, I think that changing the output order should be the exception and not the rule

I agree that silently changing the output order is generally not a good idea. Here, however, it’s part of the "prettification" meant for human consumption on the terminal, and the HTTP header and JSON key order is insignificant. Also, this behaviour has been the default for a long time, so a lot of people would have to change their config files to get back what they’re used to.

> perhaps also add a `--format-options sort=[true|false]` shortcut to set both `headers.sort` and `json.sort` at the same time
> 
> more feedback:
> I think the easiest thing to use from the command line would be to allow something like:
> ```
> --pretty=colors,format,unsorted
> --pretty=all,unsorted
> ```

Good point, a shortcut for toggling both the sortings might be handy.

Thanks for the feedback!

**Comment by jakubroztocil** [comment]
Just released v2.2.0 that addresses this issue. Learn about about the new `--unsorted`, `--sorted` and `--format-options` here:

https://httpie.org/docs#colors-and-formatting
